util/filesort/filesort.go

```diff
@@ -0,0 +1,508 @@
+
+		outputByte = append(outputByte, head...)
+		outputByte = append(outputByte, body...)
+	}
+
+	_, err = outputFile.Write(outputByte)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	fs.files = append(fs.files, fileName)
+	fs.buf = fs.buf[:0]
+	return nil
+}
+
+// Input adds one row into FileSorter.
+// Caller should not call Input after calling Output.
+func (fs *FileSorter) Input(key []types.Datum, val []types.Datum, handle int64) error {
+	if fs.closed {
+		return errors.New("FileSorter has been closed")
+	}
+	if fs.fetched {
+		return errors.New("call input after output")
+	}
+
+	// Sanity checks
+	if len(key) != fs.keySize {
+		return errors.New("mismatch in key size and key slice")
+	}
+	if len(val) != fs.valSize {
+		return errors.New("mismatch in value size and val slice")
+	}
+
+	if len(fs.buf) >= fs.bufSize {
+		err := fs.flushToFile()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	row := &comparableRow{
+		key:    key,
+		val:    val,
+		handle: handle,
+	}
+	fs.buf = append(fs.buf, row)
+	return nil
+}
+
+// Fetch the next row given the source file index.
+func (fs *FileSorter) fetchNextRow(index int) (*comparableRow, error) {
+	var (
+		err  error
+		n    int
+		head = make([]byte, 8)
+		dcod = make([]types.Datum, 0, fs.keySize+fs.valSize+1)
+	)
+	n, err = fs.fds[index].Read(head)
+	if err == io.EOF {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if n != 8 {
+		return nil, errors.New("incorrect header")
+	}
+	rowSize := int(binary.BigEndian.Uint64(head))
+
+	rowBytes := make([]byte, rowSize)
+
+	n, err = fs.fds[index].Read(rowBytes)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if n != rowSize {
+		return nil, errors.New("incorrect row")
+	}
+
+	dcod, err = codec.Decode(rowBytes, fs.keySize+fs.valSize+1)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &comparableRow{
+		key:    dcod[:fs.keySize],
+		val:    dcod[fs.keySize : fs.keySize+fs.valSize],
+		handle: dcod[fs.keySize+fs.valSize:][0].GetInt64(),
+	}, nil
+}
+
+func (fs *FileSorter) openAllFiles() error {
+	for _, fname := range fs.files {
+		fd, err := os.Open(fname)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		fs.fds = append(fs.fds, fd)
+	}
+	return nil
+}
+
+func (fs *FileSorter) closeAllFiles() error {
+	for _, fd := range fs.fds {
+		err := fd.Close()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	err := os.RemoveAll(fs.tmpDir)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Close terminates the output process and discards all remaining data.
+func (fs *FileSorter) Close() error {
+	if fs.closed {
+		return errors.New("FileSorter has been closed")
+	}
+	fs.buf = fs.buf[:0]
+	err := fs.closeAllFiles()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	fs.closed = true
+	return nil
+}
+
+// Output gets the next sorted row.
+func (fs *FileSorter) Output() (key []types.Datum, val []types.Datum, handle int64, err error) {
```

Comment:
Since the type `comparableRow` is not exported and `Output` is called by the client, I think it is more suitable to use `key`, `val` and `handle` as return types. Am I right?